### PR TITLE
Move bpfd directory/file management out of setup.sh and into code

### PR DESCRIPTION
**Pull request by Billy99**
This implements #166.

**First commit:**
From: https://www.freedesktop.org/software/systemd/man/systemd.exec.html systemd manages the following directories:
* ConfigurationDirectory: `/etc/`
* RuntimeDirectory: `/run/`  Note: RuntimeDirectory is deleted when service stops, unless overwritten by `RuntimeDirectoryPreserve`.
* CacheDirectory: `/var/cache/`
* StateDirectory: `/var/lib/`
* LogsDirectory:  `/var/log/`

Based on that, made the following changes:
 * Moved all directories from `/var/run/` to `/run/`. On most systems, `/var/run/` is a symlink to `/run/`, but there are systems where that is not the case, so made sure bpfd directories are always in `/run`.
 * Moved bytecode from `/var/bpfd/bytecode/` to the RuntimeDirectory `/run/bpfd/bytecode/` since `/var/bpfd` is not a directory managed by systemd.
 * Moved the socket used by gocounter from the ConfigurationDirectory `/etc/bpfd/sock/` to the StateDirectory `/var/lib/bpfd/sock`.
 * Added a new file, `bpfd-api/src/util.rs` and added all the directory strings. All hardcoded strings in the code were replaced the const strings.
 * Updated the `bpfd.service` file to create all the directories and set the mode such that files created under the directory inherit the user group. Also gave users within the same user group RO access to the directories, same as the script was doing. (mode=6755)
 * Updated the scripts to remove all the directory management (except certs, which will come in a follow-up PR).
 * Updated bpfd to create all the directories in one spot for the case where bpfd is not run as a systemd service.

**Second Commit:**
Create CA Certificate and Key, and bpfd Certificate and Key if they do not exist from within bpfd. Used https://github.com/sfackler/rust-openssl/blob/master/openssl/examples/mk_certs.rs to determine the logic to create certificates in rust.

**Third Commit:**
Add a oneshot bpfctl service that will create the needed directories (`/etc/bpfctl/certs/bpfctl`) as the user:group bpfctl:bpfd, then creates the required bpfctl certs.

This adds a new `bpfctl certs` command that creates the certs if they don't exist. If certs doesn't exist when other commands are entered (like `bpfctl list`), the commands fail.

The `./scripts/setup.sh install` still copies the binaries over. Then a command was added, `sudo systemctl enable bpfctl`. Then when the bpfd service is started, bpdctl service will be started by a `WantedBy=bpfd.service` line in the bpfctl.service file.

If certs need to be regenerated, remove the existing certs files and restart the serivces (bpfd must be restarted first):
```
sudo rm -rf /etc/bpfd/certs/
sudo rm -rf /etc/bpfctl/certs/
sudo systemctl restart bpfd
sudo systemctl restart bpfctl
```

The setup.sh script now only needs to support:
```
  ./scripts/setup.sh install
  ./scripts/setup.sh reinstall
  ./scripts/setup.sh uninstall
  ./scripts/setup.sh gocounter
```

Signed-off-by: Billy McFall <22157057+Billy99@users.noreply.github.com>

### CODE REVIEW

File: bpfd-api/src/util.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    //
+    // RuntimeDirectory: /run/bpfd/
+    pub const RTDIR:            &str = "/run/bpfd";
+    pub const RTDIR_BYTECODE:   &str = "/run/bpfd/bytecode";
+    pub const RTDIR_DISPATCHER: &str = "/run/bpfd/dispatchers";
+    pub const RTDIR_FS:         &str = "/run/bpfd/fs";
+    pub const RTDIR_FS_MAPS:    &str = "/run/bpfd/fs/maps";
+    pub const RTDIR_PROGRAMS:   &str = "/run/bpfd/programs";
+    //
+    // CacheDirectory: /var/cache/
```
Comment (astoycos) [inline bpfd-api/src/util.rs:24]: Is this an erroneous comment? 

Same with `LogsDirectory`

Comment (astoycos) [inline bpfd-api/src/util.rs:24]: Were the `CacheDirectory` and `LogsDirectory` Vars left here by accident?

File: bpfd/src/main.rs [hunk 2]
```diff
@@ -53,23 +51,31 @@ fn main() -> anyhow::Result<()> {
             );
             setrlimit(Resource::RLIMIT_MEMLOCK, RLIM_INFINITY, RLIM_INFINITY).unwrap();
 
-            create_dir_all(BPFFS).context("unable to create mountpoint")?;
-            create_dir_all("/var/run/bpfd/programs")?;
-            create_dir_all("/var/run/bpfd/dispatchers")?;
+            create_dir_all(RTDIR)?;
+            create_dir_all(RTDIR_FS).context("unable to create mountpoint")?;
```
Comment (astoycos) [inline bpfd/src/main.rs:55]: Any reason for only adding context to one of these create calls?

File: bpfd-api/src/util.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    // via the bpfd.service settings. They will be manually created below in case bpfd
+    // is not being run as a service.
+    //
+    // ConfigurationDirectory: /etc/bpfd/
+    pub const CFGDIR:                 &str = "/etc/bpfd";
+    pub const CFGDIR_CA_CERTS:        &str = "/etc/bpfd/certs/ca";
+    pub const CFGDIR_STATIC_PROGRAMS: &str = "/etc/bpfd/programs.d";
+    pub const CFGPATH_BPFD_CONFIG:    &str = "/etc/bpfd/bpfd.toml";
+    //
+    // RuntimeDirectory: /run/bpfd/
```
Comment (astoycos) [inline bpfd-api/src/util.rs:16]: What was the reason to move from `/var/run` again?

**Comment by astoycos** [comment]
This generally looks good to me but should probably get a pass from @dave-tucker Since I'm not super experienced as to where things _should_ be living in root

File: bpfd-api/src/util.rs (hunk 1 above)
Comment (Billy99) [reply bpfd-api/src/util.rs:24]: I left it in to spark conversation. These are the directories managed by systemd. Do some of the current runtime directories really belong in the CacheDirectory? By default, RunTime is deleted when daemon is stopped. Overriding with `RuntimeDirectoryPreserve=yes`.

Do we want logs in `/var/log/bpfd` or just with all the other logs?

Comment (Billy99) [reply bpfd-api/src/util.rs:24]: One comment wasn't enough?

File: bpfd/src/main.rs (hunk 2 above)
Comment (Billy99) [reply bpfd/src/main.rs:55]: That's the way the existing code was, I just added more directories and moved them behind const strings.

File: bpfd-api/src/util.rs (hunk 3 above)
Comment (Billy99) [reply bpfd-api/src/util.rs:16]: Systemd manages `/run/`, not `/var/run/`. On most systems, `/var/run/` is a symlink to `/run/` so it is a noop, but this is not guaranteed.

File: bpfd-api/src/util.rs (hunk 1 above)
Comment (astoycos) [reply bpfd-api/src/util.rs:24]: GH glitch lol

File: bpfd/src/main.rs (hunk 2 above)
Comment (astoycos) [reply bpfd/src/main.rs:55]: Wanna just go and add context to all of them it could be a nicer error message if ever encountered :)

**Review by dave-tucker — APPROVED** [review]
LGTM - this looks awesome! Thanks @Billy99

**Pull request merged**